File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])

        def dfs(i, j):
            if 0 <= i < m and 0 <= j < n and grid[i][j] == "1":
                grid[i][j] = 0
                dfs(i + 1, j)
                dfs(i, j + 1)
                dfs(i - 1, j)
                dfs(i, j - 1)
        
        result = 0
        for x in range(m):
            for y in range(n):
                if grid[x][y] == "1": 
                    result += 1
                    dfs(x, y)
        
        return result

File: 0200-number-of-islands/0200-number-of-islands.c
int numIslands(char** grid, int gridSize, int* gridColSize) {
    int m = gridSize, n = *gridColSize;

    void dfs(int i, int j) {
        if (0 <= i && i < m && 0 <= j && j < n && grid[i][j] == '1') {
            grid[i][j] = '0';
            dfs(i + 1, j);
            dfs(i, j + 1);
            dfs(i - 1, j);
            dfs(i, j - 1);
        }
    }

    int result = 0;
    for (int x = 0; x < m; x++) {
        for (int y = 0; y < n; y++) {
            if (grid[x][y] == '1') {
                result += 1;
                dfs(x, y);
            }
        }
    }

    return result;
}
